### DISCUSSION TITLE
IntoTempTableAsync or CreateTempTable with identity

### DISCUSSION THREAD

**Discussion by MihailsKuzmins** [linq2db/linq2db#3792] [Q&A] [answered]
Basically I have a problem with inserting my model (which has the Identity PK) into a temp table. The problem is that the Identity column is also generated for the temp table. Is there a way to disable Identity generation only for temp tables?

My model has many properties and I need to select all of them. Creating a new model with the same properties would be error-prone. So I would be grateful for some pieces of advice.

Here is the model
```cs
[Table("tblTable")]
public sealed record MyRecord
{
    [Column("PK", IsPrimaryKey = true, IsIdentity = true)]
    public int Id { get; init; }

    [Column("Value", CanBeNull = false)]
    public string Value { get; init; } = string.Empty;
}
```

Here is the code
```cs
await using var connection = OpenConnection();

connection.MyModel
    .Where(x => true)
    .IntoTempTableAsync();
```

Here is the generated SQL
```sql
CREATE TABLE [tempdb]..[#tblTable]
(
	[PK]      Int             NOT NULL IDENTITY,
	[Value]   NVarChar(4000)  NOT NULL,

	PRIMARY KEY CLUSTERED ([OK])
)
```

Here is the error
Cannot insert explicit value for identity column in table '#tblTable' when IDENTITY_INSERT is set to OFF

**Comment by MaceWindu** [comment]
As `IntoTempTableAsync` use `BulkCopy` functionality internally, it should work if you will pass `options: new BulkCopyOptions() { KeepIdentity = true }` parameter to it

**Comment by MihailsKuzmins** [comment]
Thanks for the suggestion, but unfortunately I cannot make it work, there is no override without IDataContext (i.e. this is not possible `IntoTempTableAsync( options: new BulkCopyOptions { KeepIdentity = true }, cancellationToken: ct)`), therefore, I pass my database connection as IDataContext (as shown below) , but in this case the command just gets stuck and is never completed (SQL time out is thrown).

New call. Is it possible to say from this code what I probably did wrong here? Maybe you need a solution where it can be reproduced?
```cs
await using var connection = OpenConnection();

await connection.MyModel
    .Where(x => true)
    .IntoTempTableAsync(connection, options: new BulkCopyOptions { KeepIdentity = true }, cancellationToken: ct);
```

And it looks like the generated SQL is still the same with the Identity ID column. Unfortunately, I was not able to check whether it works or not because the command never completed for me :(

Of course I came up with a dirty workaround for now, but I would like to go with the temp table
```cs
await using var connection = OpenConnection();

var items = await connection.MyModel
    .Where(x => true)
    .ToArrayAsync();

await using var tmpTable = await connection.CreateTempTableAsync(items.Select(static x => new { x.Id }), cancellationToken: ct);
```

**Reply by MaceWindu** [reply, marked as answer]
I see. I've added feature request #3795

As workaround I would recommend to split table creation and insert into two operations explicitly:
```cs
// 1. call CreateTempTableAsync to create temporary table
await using var tmp = db.CreateTempTableAsync<Model>(cancellationToken: ct);
// 2. use insert from IQueryable source
// from InsertAsync code I see it ignores Identity flag (which is strange actually), so it should work
await tmp.InsertAsync(db.MyModel, ct);
```
